tokio/src/sync/batch_semaphore.rs

```diff
@@ -0,0 +1,681 @@
+            };
+
+            if remaining > 0 && lock.is_none() {
+                // No permits were immediately available, so this permit will
+                // (probably) need to wait. We'll need to acquire a lock on the
+                // wait queue.
+                if let Ok(l) = self.waiters.try_lock() {
+                    // If we were able to acquire the lock *now* without
+                    // blocking, we don't need to reload the semaphore's permit state.
+                    lock = Some(l)
+                } else {
+                    lock = Some(self.waiters.lock().unwrap());
+                    // While we were waiting to lock the wait list, additional
+                    // permits may have been released. Therefore, we will acquire a
+                    // new snapshot of the current state of the semaphore before
+                    // actually enqueueing the waiter..
+                    curr = self.permits.load(Acquire);
+                    continue;
+                }
+            }
+
+            match self
+                .permits
+                .compare_exchange_weak(curr, next, AcqRel, Acquire)
+            {
+                Ok(_) => {
+                    acquired += acq;
+                    if remaining == 0 && !queued {
+                        return Ready(Ok(()));
+                    }
+                    break lock.unwrap();
+                }
+                Err(actual) => curr = actual,
+            }
+            drop(lock.take());
+        };
+
+        if node.assign_permits(&mut acquired) {
+            if acquired > 0 {
+                // We ended up with more permits than we needed. Release the
+                // back to the semaphore.
+                self.add_permits_locked(acquired, waiters);
+            }
+            return Ready(Ok(()));
+        }
+
+        assert_eq!(acquired, 0);
+
+        // Otherwise, register the waker & enqueue the node.
+        unsafe {
+            node.waker.with_mut(|waker| {
+                // Safety: the wait list is locked, so we may modify the waker.
+                *waker = Some(cx.waker().clone());
+            });
+
+            let node = Pin::into_inner_unchecked(node) as *mut _;
+            let node = NonNull::new_unchecked(node);
+
+            // If the waiter is not already in the wait queue, enqueue it.
+            if !waiters.contains(node) {
+                waiters.queue.push_front(node);
+            }
+        }
+
+        Pending
+    }
+}
+
+impl Drop for Semaphore {
+    fn drop(&mut self) {
+        self.close();
+    }
+}
+
+impl fmt::Debug for Semaphore {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt.debug_struct("Semaphore")
+            .field("permits", &self.permits.load(Relaxed))
+            .finish()
+    }
+}
+
+impl Permit {
+    /// Creates a new `Permit`.
+    ///
+    /// The permit begins in the "unacquired" state.
+    pub(crate) fn new() -> Permit {
```

Comment:
Having `Permit::new` as a separate API was a work around to deal w/ the fact that `Permit` allocated. I don't think this should be needed anymore. Instead, `Semaphore` can just have an `async fn acquire(...) -> Permit` ? Then, `Permit` should just become:

```struct
struct Permit {
    num: u16,
}
```

Reply:
You may consider completely getting rid of `Permit` as a struct.

Reply:
Done in 2d8d413!